core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -303,4 +322,157 @@ public boolean isSnapshotting() {
+     */
+    public BufferedJournalApplier(Map<String, RaftJournal> journals,
+        Supplier<Set<JournalSink>> journalSinks) {
+      mJournals = journals;
+      mJournalSinks = journalSinks;
+    }
+
+    /**
+     * Processes given journal entry for applying.
+     * Entry could be applied or buffered based on buffer's state.
+     *
+     * TODO(ggezer): Make this method concurrent to avoid {@link #resume} blocking quorum updates.
+     *
+     * @param journalEntry the journal entry
+     */
+    public synchronized void processJournalEntry(JournalEntry journalEntry) {
+      if (!mSuspended || journalEntry.getSequenceNumber() <= mAdvanceSequence) {
+        applyToMaster(journalEntry);
+      } else {
+        mSuspendBuffer.add(journalEntry);
+      }
+    }
+
+    /**
+     * @return {@code true} if this applier was suspended
+     */
+    public synchronized boolean isSuspended() {
+      return mSuspended;
+    }
+
+    /**
+     * Suspend the applier.
+     *
+     * Subsequent journal entries will be buffered until
+     * {@link #resume()} or {@link #advance(long)} is called.
+     *
+     * @throws IOException
+     */
+    public synchronized void suspend() throws IOException {
+      Preconditions.checkState(!mSuspended, "Already suspended");
+      LOG.info("Suspended state machine at sequence: {}", mLastAppliedSequence);
+      mSuspended = true;
+      mAdvanceSequence = -1;
+      mAdvanceFuture = null;
+    }
+
+    /**
+     * Resumes the applier.
+     * Before returning this method will apply all buffered entries.
+     *
+     * @throws IOException
+     */
+    public synchronized void resume() throws IOException {
+      Preconditions.checkState(mSuspended, "Not suspended");
+      LOG.info("Resuming state machine.");
+      try {
+        for (JournalEntry entry : mSuspendBuffer) {
+          applyToMaster(entry);
+        }
+        mSuspendBuffer.clear();
+      } finally {
+        mSuspended = false;
+        mAdvanceSequence = -1;
+        mAdvanceFuture = null;
+      }
+    }
+
+    /**
+     * Advances the applier to a target sequence.
+     * This method leaves the applier in suspended state.
+     *
+     * @param sequence target sequence
+     * @return the future when applier reaches the target sequence
+     */
+    public synchronized CompletableFuture<Void> advance(long sequence) {
+      Preconditions.checkState(sequence >= 0, "Invalid sequence");
+      Preconditions.checkState(mSuspended, "Not suspended");
+      Preconditions.checkState(mAdvanceFuture == null, "Advance future already set");
+      LOG.info("Advancing state machine to sequence: {}", sequence);
+      mAdvanceSequence = sequence;
+      // Complete the request if already at target sequence.
+      if (mLastAppliedSequence == mAdvanceSequence) {
+        return CompletableFuture.completedFuture(null);
+      }
+      // Create a future for completion.
+      mAdvanceFuture = new CompletableFuture<>();
```

Comment:
Looks like the advance behavior is synchronous, 
it will always be   
```
mAdvanceFuture = new CompletableFuture<>();
while loop to applyToMaster() -> mAdvanceFuture.complete(null);
return advanceFuture
```
Why do we need to use a completableFuture?

Reply:
It's synchronous only for the part that is in the buffer. If target sequence is not met, then future will not be completed until state machine receives it.

Reply:
Got it! Thanks